src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -107,13 +107,79 @@ bool RaftReplDev::join_group() {
         m_msg_mgr.join_group(m_group_id, "homestore_replication",
                              std::dynamic_pointer_cast< nuraft_mesg::mesg_state_mgr >(shared_from_this()));
     if (!raft_result) {
-        HS_DBG_ASSERT(false, "Unable to join the group_id={} with error={}", boost::uuids::to_string(m_group_id),
-                      raft_result.error());
+        HS_DBG_ASSERT(false, "Unable to join the group_id={} with error={}", group_id_str(), raft_result.error());
         return false;
     }
     return true;
 }
 
+AsyncReplResult<> RaftReplDev::replace_member(replica_id_t member_out_uuid, replica_id_t member_in_uuid) {
+    LOGINFO("Replace member group_id={} member_out={} member_in={}", group_id_str(),
+            boost::uuids::to_string(member_out_uuid), boost::uuids::to_string(member_in_uuid));
+
+    // Step 1: Check if leader itself is requested to move out.
+    if (m_my_repl_id == member_out_uuid && m_my_repl_id == get_leader_id()) {
+        // If leader is the member requested to move out, then give up leadership and return error.
+        // Client will retry replace_member request to the new leader.
+        raft_server()->yield_leadership(true /* immediate */, -1 /* successor */);
+        LOGINFO("Replace member leader is the member_out so yield leadership");
+        return make_async_error<>(ReplServiceError::NOT_LEADER);
+    }
+
+    // Step 2. Add the new member.
+    return m_msg_mgr.add_member(m_group_id, member_in_uuid)
+        .via(&folly::InlineExecutor::instance())
+        .thenValue([this, member_in_uuid, member_out_uuid](auto&& e) -> AsyncReplResult<> {
+            // TODO Currently we ignore the cancelled, fix nuraft_mesg to not timeout
+            // when adding member. Member is added to cluster config until member syncs fully
+            // with atleast stop gap. This will take a lot of time for block or
+            // object storage.
+            if (e.hasError() && e.error() != nuraft::cmd_result_code::CANCELLED) {
+                RD_LOGE("Replace member error in add member : {}", e.error());
+                return make_async_error<>(RaftReplService::to_repl_error(e.error()));
+            }
+            auto member_out = boost::uuids::to_string(member_out_uuid);
+            auto member_in = boost::uuids::to_string(member_in_uuid);
+
+            LOGINFO("Replace member added member={} to group_id={}", member_in, group_id_str());
+
+            // Step 3. Append log entry to mark the old member is out and new member is added.
```

Comment:
if leader changes after line140 and before line 147, then who can send a log entry to mark the old member is out and remove member_out of out this group?

should we separate replace_member to two calls: add_member and remove_member?

Reply:
Yeah valid scenario. In that case log entry append will fail at line 157, CM will retry again with the latest leader. Thats the flow we have decided with CM. It should be safe to send the same message to a different leader.

Reply:
>It should be safe to send the same message to a different leader.

if leader changes after `propose_to_raft` but before `m_msg_mgr.rem_member`, then it will return `ReplServiceError::RETRY_REQUEST` .

now , if cm resends the request to the new leader, I think `add_member` will succeed in new leader since the member_in is already added in the group by the previous leader. now, next step is `propose_to_raft`, which will create another `HS_CTRL_REPLACE` log and commit it.

so `HS_CTRL_REPLACE` will be committed twice at the other follower, one is from  previous leader , one is from current leader.

we should make it safe to commit `HS_CTRL_REPLACE` twice（or make committing `HS_CTRL_REPLACE` idempotent） for the same member_out and member_in. so we need to add some checks in `m_listener#replace_member(member_out, member_in);`

so, I think basically it will be safe,  but we should add some code to handle the corner case.

Reply:
more for cm:
if the new leader is the `member_in`,   we havent respond to the call yet.   We get the leader (in nuraft_mesg) from raft which is always up_to_date,  and respond to CM.

Wondering if CM expect HB from a to-be-added SM , or expect retry aganist a `to-be-added SM`.

Reply:
idempotency of HS_CTRL_REPLACE is same as create pg or shard. Gateway can send request leader, leader does raft and before returning success to gateway crashed and restarted, gateway will send to same leader and we might create pg. In HO in add_replace function , in and out members are already applied, its a nop.